### PR TITLE
[Component Manager] Fix race condition

### PR DESCRIPTION
**Pull request by smnzhu**
closes https://github.com/dapperlabs/flow-go/issues/6220

Fixes race condition introduced in https://github.com/onflow/flow-go/pull/1543

### CODE REVIEW

**Review by peterargue — COMMENTED** [review]
can you add some notes about what the race condition is as comments (or in the PR description).

File: module/component/component.go [hunk 1]
```diff
@@ -218,14 +218,14 @@ func (c *ComponentManager) Start(parent irrecoverable.SignalerContext) {
 		// goroutine and the parent's are scheduled. If the parent is scheduled first, any errors
 		// thrown within workers would not have propagated, and it would only receive the done signal
 		defer func() {
+			cancel() // shutdown all workers
+			<-c.shutdownSignal
```
Comment (jordanschalm) [inline module/component/component.go:222]:
```suggestion
			cancel() // shutdown all workers
			// wait for shutdown signal before signalling the component is done
			// this guarantees that ShutdownSignal is closed before Done
			<-c.shutdownSignal
```

File: module/component/component_manager_test.go [hunk 2]
```diff
@@ -17,7 +17,7 @@ import (
 	"github.com/onflow/flow-go/utils/unittest"
 )
 
-const CHANNEL_CLOSE_LATENCY_ALLOWANCE = 20 * time.Millisecond
+const CHANNEL_CLOSE_LATENCY_ALLOWANCE = 25 * time.Millisecond
```
Comment (jordanschalm) [inline module/component/component_manager_test.go:20]: Should we add the test reproducing the race? It's `TestComponentManagerShutdown` in https://github.com/onflow/flow-go/commit/803b173a753dffe658c9c0695f59cff5a1f2d893

**Review by jordanschalm — COMMENTED** [review]
Thanks for the fix Simon, this looks good to me.

I tested with [the case](https://github.com/onflow/flow-go/commit/803b173a753dffe658c9c0695f59cff5a1f2d893) I was using to reproduce earlier, and it passes now. I think we should add that test case in this PR as well.

**Comment by smnzhu** [comment]
> can you add some notes about what the race condition is as comments (or in the PR description).

@peterargue See the issue referenced in the description: https://github.com/dapperlabs/flow-go/issues/6220

**Review by peterargue — APPROVED** [review]
Looks good. Probably worth including Jordan's script.

**Review by jordanschalm — APPROVED** [review]

**Comment by smnzhu** [comment]
bors merge

**Pull request merged**